storage/sftp/storage_sftp.go

```diff
@@ -0,0 +1,157 @@
+/*
+ * knoxite
+ *     Copyright (c) 2019, Fabian Siegel <fabians1999@gmail.com>
+ *
+ *   For license see LICENSE
+ */
+
+package sftp
+
+import (
+	"errors"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"os"
+	"os/user"
+	"path/filepath"
+
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+	agent "golang.org/x/crypto/ssh/agent"
+	kh "golang.org/x/crypto/ssh/knownhosts"
+
+	knoxite "github.com/knoxite/knoxite/lib"
+)
+
+type StorageSFTP struct {
+	url   url.URL
+	ssh   *ssh.Client
+	sftp  *sftp.Client
+	login bool
+	knoxite.StorageFilesystem
+}
+
+func init() {
+	knoxite.RegisterBackendFactory(&StorageSFTP{})
+}
+
+func (*StorageSFTP) NewBackend(u url.URL) (knoxite.Backend, error) {
+	_, port, err := net.SplitHostPort(u.Host)
+	if err != nil || len(port) == 0 {
+		port = "22"
+		u.Host = net.JoinHostPort(u.Host, port)
+	}
+	username := u.User.Username()
+	password, isSet := u.User.Password()
+
+	auth := []ssh.AuthMethod{}
+	if isSet {
+		auth = append(auth, ssh.Password(password))
+	} else {
+		socket := os.Getenv("SSH_AUTH_SOCK")
+		agent_conn, err := net.Dial("unix", socket)
+		if err != nil {
+			return &StorageSFTP{}, knoxite.ErrInvalidPassword
+		} else {
+			agentClient := agent.NewClient(agent_conn)
+			auth = append(auth, ssh.PublicKeysCallback(agentClient.Signers))
+		}
+	}
+
+	usr, _ := user.Current()
+
+	hostKeyCallback, err := kh.New(filepath.Join(usr.HomeDir, ".ssh/known_hosts"))
+	if err != nil {
+		// If no hostkey can be found, ignore it for now...
+		hostKeyCallback = ssh.InsecureIgnoreHostKey()
+	}
+
+	config := &ssh.ClientConfig{
+		User:            username,
+		Auth:            auth,
+		HostKeyCallback: hostKeyCallback,
+	}
+
+	conn, err := ssh.Dial("tcp", u.Hostname()+":"+u.Port(), config)
+	if err != nil {
+		panic("Failed to dial: " + err.Error())
+	}
+
+	client, err := sftp.NewClient(conn)
+	if err != nil {
+		panic("Failed to create client: " + err.Error())
```

Comment:
Don't panic here. Simply return the error you encountered and let knoxite handle informing the user.